internal/messagejournal/messagejournal.go

```diff
@@ -0,0 +1,231 @@
+	if err = db.Ping(); err != nil {
+		return nil, fmt.Errorf("message journal database not connected: %w", err)
+	}
+
+	return &messageJournal, nil
+}
+
+// migrateMessageJournalDB handles the migration of the message journal
+// database and ensures the schema is up to date on each session start.
+func migrateMessageJournalDB(db *sql.DB, databaseFilePath string) error {
+	databaseDriver, err := sqlite.WithInstance(db, &sqlite.Config{})
+	if err != nil {
+		return fmt.Errorf("database driver not initialized: %w", err)
+	}
+	migrationDriver, err := iofs.New(embeddedMigrationData, "migrations")
+	if err != nil {
+		return fmt.Errorf("embedded migration data not found: %w", err)
+	}
+	migration, err := migrate.NewWithInstance(
+		"iofs",
+		migrationDriver,
+		databaseFilePath,
+		databaseDriver,
+	)
+	if err != nil {
+		return fmt.Errorf("database migration not initialized: %w", err)
+	}
+	if err = migration.Up(); err != nil && err != migrate.ErrNoChange {
+		return fmt.Errorf("database migration failed: %w", err)
+	}
+	return nil
+}
+
+// AddEntry adds a new message journal entry to both the temporary runtime table
+// which gets cleared at the start of every session and the persistent table
+// which maintains message entries across multiple sessions.
+func (j *MessageJournal) AddEntry(entry yggdrasil.WorkerMessage) (*yggdrasil.WorkerMessage, error) {
+	const insertEntryTemplate string = `INSERT INTO %s (
+		message_id, sent, worker_name, response_to, worker_event, worker_message) 
+		values (?,?,?,?,?,?)`
+
+	insertAction, err := j.database.Prepare(fmt.Sprintf(insertEntryTemplate, messageJournalTableName))
+	if err != nil {
+		return nil, fmt.Errorf("cannot prepare statement for table '%v': %w", messageJournalTableName, err)
+	}
+
+	persistentResult, err := insertAction.Exec(
+		entry.MessageID,
+		entry.Sent,
+		entry.WorkerName,
+		entry.ResponseTo,
+		entry.WorkerEvent.EventName,
+		entry.WorkerEvent.EventMessage,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not insert journal entry into table '%v': %w", messageJournalTableName, err)
+	}
+
+	entryID, err := persistentResult.LastInsertId()
+	if err != nil {
+		return nil, fmt.Errorf("could not select last insert ID '%v' for table '%v': %w", entryID, messageJournalTableName, err)
+	}
+	log.Debugf("new message journal entry (id: %v) added: '%v'", entryID, entry.MessageID)
+
+	return &entry, nil
+}
+
+// GetEntries retrieves a list of all the journal entries in the message journal database
+// that meet the criteria of the provided message journal filter.
+func (j *MessageJournal) GetEntries(filter Filter) ([]map[string]string, error) {
+	entries := []map[string]string{}
+	queryString, err := j.buildDynamicGetEntriesQuery(filter)
+	if err != nil {
+		return nil, fmt.Errorf("cannot build dynamic sql query: %w", err)
+	}
+
+	preparedQuery, err := j.database.Prepare(queryString)
+	if err != nil {
+		return nil, fmt.Errorf("cannot prepare query when retrieving journal entries: %w", err)
```

Comment:
Almost-but-not-quite matching error string text. I like "to retrieve" more than
"when retrieving" personally.

```suggestion
		return nil, fmt.Errorf("cannot prepare query to retrieve journal entries: %w", err)
```